src/client/render/secondstage.cpp

```diff
@@ -0,0 +1,100 @@
+/*
+Minetest
+Copyright (C) 2010-2013 celeron55, Perttu Ahola <celeron55@gmail.com>
+Copyright (C) 2017 numzero, Lobachevskiy Vitaliy <numzer0@yandex.ru>
+Copyright (C) 2020 appgurueu, Lars Mueller <appgurulars@gmx.de>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "secondstage.h"
+#include "client/client.h"
+#include "client/shader.h"
+#include "client/tile.h"
+
+RenderingCoreSecondStage::RenderingCoreSecondStage(
+		IrrlichtDevice *_device, Client *_client, Hud *_hud) :
+		RenderingCoreStereo(_device, _client, _hud)
+{
+	initMaterial();
+}
+
+void RenderingCoreSecondStage::initMaterial()
+{
+	IWritableShaderSource *s = client->getShaderSource();
+	s->global_additional_headers = "#define SECONDSTAGE 1\n";
+	s->rebuildShaders(); // TODO remove this if possible, use shader const setters?
+	mat.UseMipMaps = false;
+	mat.ZBuffer = true;
+	mat.ZWriteEnable = true;
+	u32 shader = s->getShader("3d_secondstage", TILE_MATERIAL_BASIC, 0);
+	mat.MaterialType = s->getShaderInfo(shader).material;
+	for (int k = 0; k < 3; ++k) {
+		mat.TextureLayer[k].AnisotropicFilter = false;
+		mat.TextureLayer[k].BilinearFilter = false;
+		mat.TextureLayer[k].TrilinearFilter = false;
+		mat.TextureLayer[k].TextureWrapU = video::ETC_CLAMP_TO_EDGE;
+		mat.TextureLayer[k].TextureWrapV = video::ETC_CLAMP_TO_EDGE;
+	}
+}
+
+void RenderingCoreSecondStage::initTextures()
+{
+	rendered = driver->addRenderTargetTexture(
+			screensize, "3d_render", video::ECF_A8R8G8B8);
+	normalmap = driver->addRenderTargetTexture(
+			screensize, "3d_normalmap", video::ECF_A8R8G8B8);
```

Comment:
Why not `ECF_A16B16G16R16F`? (is supported in the IrrLicht version MT uses?)

Reply:
I have read somewhere that it was best for performance if all output textures had the same "bytewidth" (all 4 bytes). Also, I think one byte is enough for normals?

Reply:
I’d prefer `GL_RGB8_SNORM` (or RGBA) probably but it’s not supported by IrrLicht. Not sure of performance, it’s to be measured.

Reply:
8 bit integer precision for the colour values are not sufficient for postprocessing in my opinion. I assume that if the colours are quantized to 8 bit and the second stage shader increases the contrast (or does something similar), [banding artifacts](https://en.wikipedia.org/wiki/Colour_banding) could appear.
As far as I know, it is also important that the colours are not clamped into the [0, 1] range to allow a High Dynamic Range.